### PR TITLE
Add compose version check

### PR DESCRIPTION
**Pull request by allisaurus**
"make" & "make test" pass

### CODE REVIEW

**Comment by SoManyHs** [comment]
Can you add what testing you did?

File: ecs-cli/modules/cli/compose/project/project.go [hunk 1]
```diff
@@ -134,11 +134,25 @@ func (p *ecsProject) Parse() error {
...
 func (p *ecsProject) parseCompose() error {
 	logrus.Debug("Parsing the compose yaml...")
+
+	// check for Compose version and call appropriate parsing function
+	version, err := p.checkComposeVersion()
+	if err != nil {
+		return err
+	}
+	switch version {
+	case "", "1", "1.0", "2", "2.0":
```
Comment (SoManyHs) [inline ecs-cli/modules/cli/compose/project/project.go]: You might consider using `strconv.ParseFloat` instead of doing a string match here, eg:
```
	v, _ := strconv.ParseFloat(version, 32)
	if v == 1 || v == 2 {
		parseV1V2
	}
```

Or something like that.

File: ecs-cli/modules/cli/compose/project/project_parseV1V2.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+import (
+	"github.com/docker/libcompose/project"
+	"github.com/sirupsen/logrus"
+)
+
+func (p *ecsProject) parseV1V2() {
+	libProject := project.NewProject(&p.ecsContext.Context, nil, nil)
+	libProject.Parse()
+
+	logrus.Debug("Parsing v1/2 project...")
```
Comment (SoManyHs) [inline ecs-cli/modules/cli/compose/project/project_parseV1V2.go]: Total nit, but since we have this debug log statement in both `parseV1V2` and `parseV3`, would it make sense to move it into the `parseCompose` function itself and have something like `logrus.Debugf("Parsing docker compose version: %v", version)`?

File: ecs-cli/modules/cli/compose/project/project_parseV3.go [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package project
+
+import "github.com/sirupsen/logrus"
+
+func (p *ecsProject) parseV3() {
+	// TODO: parse v3, convert ServiceConfigs to ContainerConfigs
+	logrus.Debug("Parsing v3 project...")
```
Comment (SoManyHs) [inline ecs-cli/modules/cli/compose/project/project_parseV3.go]: See above

File: ecs-cli/modules/cli/compose/project/project_version.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+)
+
+func (p *ecsProject) checkComposeVersion() (string, error) {
+	var composeVersion string
+	for _, file := range p.ecsContext.ComposeFiles {
+		fileVersion, err := getFileVersion(file)
+		if err != nil {
+			return "", err
+		}
+		if len(p.ecsContext.ComposeFiles) > 0 && composeVersion != "" && composeVersion != fileVersion {
```
Comment (SoManyHs) [inline ecs-cli/modules/cli/compose/project/project_version.go]: Given that most customers use only one compose file, could we maybe check if `len(p.ecsContext.ComposeFiles) == 1` and return the version from `getFileVersion` early (or return an error saying there needs to be at least one ComposeFile present). Then if we do need to iterate over multiple files, that would allow us to simplify this conditional to just checking `composeVersion != "" && composeVersion != fileVersion`, I believe.

File: ecs-cli/modules/cli/compose/project/project_version.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+		composeVersion = fileVersion
+	}
+	return composeVersion, nil
+}
+
+func getFileVersion(file string) (string, error) {
+	type ComposeVersion struct {
+		Version string `json:"version"`
+	}
+	version := &ComposeVersion{}
```
Comment (SoManyHs) [inline ecs-cli/modules/cli/compose/project/project_version.go]: Hm, seems a little odd to have the type defined in the function itself. Are you thinking this won't be used anywhere else?

File: ecs-cli/modules/cli/compose/project/project.go (hunk 1 above)
Comment (PettitWesley) [reply ecs-cli/modules/cli/compose/project/project.go]: Why not just check if the string starts with "2" or "3"?

Comment (PettitWesley) [reply ecs-cli/modules/cli/compose/project/project.go]: If you hard code the exact strings, then we are stuck supporting those version strings only (related: https://github.com/aws/amazon-ecs-cli/issues/305).

Comment (PettitWesley) [reply ecs-cli/modules/cli/compose/project/project.go]: I like @SoManyHs idea of parsing into a float, but does Compose use semantic versioning? "2.1.5" is a valid version but not a valid float.

Comment (SoManyHs) [reply ecs-cli/modules/cli/compose/project/project.go]: Looks like compose using "x.x" versioning: https://docs.docker.com/compose/compose-file/compose-versioning/#compatibility-matrix, so `strconv` might still be alright. I am wondering though if we were only accounting for 3.0 vs the minor versions of compose 3, since there do appear to be other fields.

File: ecs-cli/modules/cli/compose/project/project_version.go (hunk 5 above)
Comment (allisaurus) [reply ecs-cli/modules/cli/compose/project/project_version.go]: yes, my thought was this is just an expedient way of extracting the version using yaml.Unmarshal(). Outside of this function the version will be handled as a string or float (pending the discussion above)

File: ecs-cli/modules/cli/compose/project/project.go (hunk 1 above)
Comment (allisaurus) [reply ecs-cli/modules/cli/compose/project/project.go]: re: versioning - Currently, libcompose throws an unmarshalling error if minor versions are detected (e.g., 2.1), so scoping 3 support the major version "3" or "3.0" would be consistent (similarly, Kompose does not support minor versions of 3). 

However, I'm all for supporting up to what is feasible up to 3.6 (latest). I suggest we start with supporting 3/3.0 and work on supporting minor versions of 3 in a future iteration(s). Thoughts?

Comment (PettitWesley) [reply ecs-cli/modules/cli/compose/project/project.go]: Why not just build it to support all versions now? Libcompose could fix their stuff so that minor versions can be unmarshalled, and if we support minor versions too, then we aren't leaving around any future work that needs to be done.

Comment (SoManyHs) [reply ecs-cli/modules/cli/compose/project/project.go]: For now, scoping the parsing to support just the major versions seems most feasible. We've already learned we can't rely on libcompose for support of different versions, and to date it doesn't seem like we've had many customers ask for minor versions of v2. As I understand it, the ask by customers for v3 was just for major versions. Limiting the scope to major releases is probably our best bet for moving forward on this.

Based on that, the original implementation should be fine. My suggestion for using `strconv` was mostly to consider edge cases and minor versions, but the latter may now be irrelevant. Including unit tests will also help with this :)

File: ecs-cli/modules/cli/compose/project/project_version.go (hunk 4 above)
Comment (SoManyHs) [reply ecs-cli/modules/cli/compose/project/project_version.go]: Sorry, perhaps I wasn't clear. I was suggesting having the check for a single compose file outside the loop entirely, and only iterate if there are multiple files. E.g:
```
// For single compose file
if len(p.ecsContext.ComposeFiles) == 1 {
        composeVersion, err := getFileVersion(file)
		if err != nil {
			return "", err
		}
        return composeVersion
}

// For multiple compose files (rare, i think?)
var composeVersion string
	for _, file := range p.ecsContext.ComposeFiles {
		fileVersion, err := getFileVersion(file)
		if err != nil {
			return "", err
		}
		if composeVersion != "" && composeVersion != fileVersion {
			logrus.Errorf("Compose files must be of the same version. Found: %s and %s", composeVersion, fileVersion)
		}
		composeVersion = fileVersion
	}
	return composeVersion, nil

File: ecs-cli/modules/cli/compose/project/project.go (hunk 1 above)
Comment (allisaurus) [reply ecs-cli/modules/cli/compose/project/project.go]: @PettitWesley my comment mainly pertains to the scope of this code review. The work needed to differentiate between features supported in different minor versions of 3 will be better understood after we start implementing parsing and task def conversion, which hasn't been started yet. I'd prefer to get basic support for v3 working and then work on supporting minor versions after we better understand what that would entail.

+1 to @SoManyHs  comment re: unit tests.

File: ecs-cli/modules/cli/compose/project/project_version.go (hunk 5 above)
Comment (SoManyHs) [reply ecs-cli/modules/cli/compose/project/project_version.go]: Yeah, I think the version should still be string regardless of how we parse it. It was just weird to see the type declaration inside the function, but if we're not using it elsewhere it should be fine.

File: ecs-cli/modules/cli/compose/project/project_version.go (hunk 4 above)
Comment (allisaurus) [reply ecs-cli/modules/cli/compose/project/project_version.go]: But why break this up into multiple cases/blocks when one loop can handle either case? Iterating through the loop once (for one file) is just as performant as doing a standalone check. Also, if we add a check for 0 files before the loop (to detect and throw an appropriate error msg) then we can simplify the check on line 18 to `composeVersion != "" && composeVersion != fileVersion`

Comment (SoManyHs) [reply ecs-cli/modules/cli/compose/project/project_version.go]: Good point! I was mostly thinking of ways to avoid 3 conditions but you're right that the checking for 0 files will take care of that. Was mostly just trying to see if there was a way to write it in a more obvious way for what's happening bc my slow brain had to think about it on the first read-through, but your implementation is 👍

File: ecs-cli/modules/cli/compose/project/project_version_test.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+	tmpfile2.Close()
+
+	// setup project and check that error is thrown for mismatches file versions
+	project := setupTestProject(t)
+	project.ecsContext.ComposeFiles = append(project.ecsContext.ComposeFiles, tmpfile1.Name())
+	project.ecsContext.ComposeFiles = append(project.ecsContext.ComposeFiles, tmpfile2.Name())
+	_, error := project.checkComposeVersion()
+
+	assert.Error(t, error)
+}
```
Comment (SoManyHs) [inline ecs-cli/modules/cli/compose/project/project_version_test.go]: Generally fine, but I'd add a test for at least:
1) an empty version string
2) two files with the same version but formatted differently (2.0 and 2)

May vary depending on how you decide to return the version type (see previous comment)

File: ecs-cli/modules/cli/compose/project/project_version.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+	yaml "gopkg.in/yaml.v2"
+)
+
+func (p *ecsProject) checkComposeVersion() (string, error) {
+	var composeVersion string
+	if len(p.ecsContext.ComposeFiles) == 0 {
+		return "", fmt.Errorf("No Compose files found")
+	}
+	for _, file := range p.ecsContext.ComposeFiles {
+		fileVersion, err := getFileVersion(file)
```
[the remark this answers is not in the document]
Comment (allisaurus) [reply ecs-cli/modules/cli/compose/project/project_version.go]: Yes, it will. As discussed offline:
 - while '2' and '2.0' represent the same version, current behavior w/ libcompose throws an error if "x.0" is detected. See https://github.com/docker/libcompose/issues/490
 - so, if "x.0" is detected, we can either 1) accept it and proceed, knowing it will error on libcompose parse(), or 2) accept it and modify the compose bytes so that '2' is passed in and accepted by libcompose
 - given the above, we are opting to 1) keep the consistency check since it will write out the detected versions, making the 2/2.0 discrepancy known 2) not modify the bytes passed to libcompose, but log out a warning that x.0 versions are not valid. This way the error will surface from libcompose (where the problem originates) but we provide better messaging re: the cause

File: ecs-cli/modules/cli/compose/project/project_parseV1V2.go [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+import (
+	"github.com/aws/amazon-ecs-cli/ecs-cli/modules/cli/compose/containerconfig"
+	"github.com/docker/libcompose/project"
+	"github.com/sirupsen/logrus"
+)
+
+func (p *ecsProject) parseV1V2() ([]containerconfig.ContainerConfig, error) {
+	libProject := project.NewProject(&p.ecsContext.Context, nil, nil)
+	libProject.Parse()
```
[the remark this answers is not in the document]
Comment (allisaurus) [reply ecs-cli/modules/cli/compose/project/project_parseV1V2.go]: good thought re: the pointer, will change.

**Comment by SoManyHs** [comment]
LGTM! 👍
